src/libraries/System.Text.Json/source_generation/src/JsonSourceGeneratorHelper.Generate.cs

```diff
@@ -0,0 +1,672 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Reflection;
+using System.Text.Json.Serialization;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Text;
+
+namespace System.Text.Json.SourceGeneration
+{
+    internal sealed partial class JsonSourceGeneratorHelper
+    {
+        private readonly string _generationNamespace;
+
+        // TODO: consider public option for this.
+        private bool _honorRuntimeProvidedCustomConverters = true;
+
+        private const string RuntimeCustomConverterFetchingMethodName = "GetRuntimeProvidedCustomConverter";
+
+        private const string JsonContextDeclarationSource = "internal partial class JsonContext : JsonSerializerContext";
+
+        private const string OptionsInstanceVariableName = "Options";
+
+        private const string PropInitFuncVarName = "PropInitFunc";
+
+        private const string JsonMetadataServicesClassName = "JsonMetadataServices";
+
+        private const string CreateValueInfoMethodName = "CreateValueInfo";
+
+        private static DiagnosticDescriptor TypeNotSupported { get; } = new DiagnosticDescriptor(
+            "SYSLIB2021",
+            "Did not generate serialization metadata for type",
```

Comment:
Perhaps for a follow-up PR: these need to be moved to resx with localization enabled.

Reply:
Added to my list of follow-ups - https://github.com/dotnet/runtimelab/projects/1#column-11545555.